src/Appwrite/Platform/Tasks/Backup.php

```diff
@@ -0,0 +1,252 @@
+            '--check-privileges', // checks if Percona XtraBackup has all the required privileges.
+            '--target-dir=' . $target,
+            '--parallel=' . self::PROCESSORS,
+            '--compress=' . self::COMPRESS_ALGORITHM,
+            '--compress-threads=' . self::PROCESSORS,
+            '--rsync', // https://docs.percona.com/percona-xtrabackup/8.0/accelerate-backup-process.html
+            //'--encrypt-threads=' . $this->processors,
+            //'--encrypt=AES256',
+            //'--encrypt-key-file=' . '/encryption_key_file',
+            //'--no-lock', // https://docs.percona.com/percona-xtrabackup/8.0/xtrabackup-option-reference.html#-no-lock
+            '2> ' . $logfile,
+        ];
+
+        $cmd = 'docker exec appwrite-xtrabackup xtrabackup ' . implode(' ', $args);
+        self::log($cmd);
+        shell_exec($cmd);
+
+        $stderr = shell_exec('tail -1 ' . $logfile);
+        Backup::log($stderr);
+
+        if (!str_contains($stderr, 'completed OK!') || !file_exists($target . '/xtrabackup_checkpoints')) {
+            Console::error('Backup failed');
+            Console::exit();
+        }
+
+        // todo: remove logfile?
+    }
+
+    public function tar(string $directory, string $file)
+    {
+        $stdout = '';
+        $stderr = '';
+        // Tar from inside the directory for not using --strip-components
+        $cmd = 'cd ' . $directory . ' && tar zcf ' . $file . ' .';
+        self::log($cmd);
+        Console::execute($cmd, '', $stdout, $stderr);
+        self::log($stdout);
+        if (!empty($stderr)) {
+            Console::error($stderr);
+            Console::exit();
+        }
+
+        if (!file_exists($file)) {
+            Console::error("Can't find tar file: " . $file);
+            Console::exit();
+        }
+
+        $filesize = \filesize($file);
+        self::log("Tar file size is: " . ceil($filesize / 1024 / 1024) . 'MB');
+        if ($filesize < (2 * 1024 * 1024)) {
+            Console::error("File size is very small: " . $file);
+            Console::exit();
+        }
+    }
+
+    public function upload(string $file, Device $local)
+    {
+        $filename = basename($file);
+        $s3 = new DOSpaces('/' . $this->database . '/full', App::getEnv('_DO_SPACES_ACCESS_KEY'), App::getEnv('_DO_SPACES_SECRET_KEY'), App::getEnv('_DO_SPACES_BUCKET_NAME'), App::getEnv('_DO_SPACES_REGION'));
+
+        if (!$s3->exists('/')) {
+            Console::error('Can\'t read s3 root directory');
+            Console::exit();
+        }
+
+        try {
+            self::log('Uploading: ' . $file);
+            $destination = $s3->getRoot() . '/' . $filename;
+
+            if (!$local->transfer($file, $destination, $s3)) {
+                Console::error('Error uploading to ' . $destination);
+                Console::exit();
+            }
+
+            if (!$s3->exists($destination)) {
+                Console::error('File not found on s3 ' . $destination);
```

Comment:
Let's use `destination` instead of `s3` so it's agnostic to the platform

Reply:
yep..makes sense 👍